### ISSUE TITLE
Feature request: tf.nn.ctc_loss lacks the API to handle sequences with all blanks

### ISSUE THREAD

**Issue by xilinli** [tensorflow/tensorflow#14659]
It is trivial to calculate the CTC loss of a sequence with all blanks. But tf.nn.ctc_loss cannot handle the situation that one or more sequences in the batch have no non-blank labels. This is a big limitation.   

P.S.: I pulled these requests before:
I reported this previously at
https://github.com/tensorflow/tensorflow/issues/13457
I asked this on stackoverflow
https://stackoverflow.com/questions/46652720/how-to-calculate-ctc-loss-of-a-sequence-with-all-blanks-using-tf-nn-ctc-loss

**Comment by cy89** [comment]
@xilinli I'm sorry we and the community haven't been able to get you an effective answer. 

@ebrevdo and @andrewharp is @xilinli doing something different with his inputs than the usual way that ctc_ops.py is used? Might there be a workaround? Or do we anticipate making things more robust for his case?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
Sounds like this is a edge case bug... Happy to review a bugfix pr.

**Comment by pinakinathc** [comment]
is anyone working on this issue or has it been fixed in a different issue or PR?
if it is still open i would like to review it and try to create a PR.

**Comment by ebrevdo** [comment]
No one is working on it afaik

**Comment by noammor** [comment]
I just hit the same bug. Has anyone found a workaround?

**Comment by pinakinathc** [comment]
Hi noammor
yes, please use the following code:
"""
>>>  y_true = tf.cast(y_true, dtype=tf.int32)
>>>  label_length = tf.multiply(tf.ones(tf.shape(y_true)[0],
dtype=tf.int32), 2)
>>>  where = tf.ones(tf.shape(y_true))
>>>  indices = tf.where(where)
>>>  values = tf.gather_nd(y_true, indices)
>>>  sparse = tf.SparseTensor(indices, values,
                            tf.cast(tf.shape(y_true), dtype=tf.int64))
  # Finally calculating the CTC Loss
>>>  loss = tf.nn.ctc_loss(sparse, y_pred, label_length, time_major=False,
                                                ctc_merge_repeated=False)
"""

hope this works for you.

**Comment by theceday** [comment]
I wonder how many people actually hitting this one without even releazing it.

**Comment by dgboy2000** [comment]
+1 Thanks for your attention. This seems to make tensorflow CTC unusable in a common use case.

**Comment by theceday** [comment]
This is fixed for pytorch. Probably similar changes needed here.
https://github.com/pytorch/pytorch/pull/23298
https://github.com/pytorch/pytorch/pull/21910
